# ĐỒ ÁN CUỐI KÌ

| MSSV     | Họ và tên       | Github    |
|----------|-----------------|-----------|
| 19120467  | Ngô Hữu Đang | ngohuudang2001 |
| 18120464 | Nguyễn Văn Minh | MinhCr007 |

---

# Import Library

In [1]:
import pandas as pd
import numpy as np
from matplotlib import pyplot as plt

---

# A. THU THẬP DỮ LIỆU

**Dữ liệu đã được lấy trong một cuộc khảo sát về các khóa học toán học và tiếng Bồ Đào Nha ở trường trung học. Nó chứa rất nhiều thông tin xã hội, giới tính và nghiên cứu thú vị về sinh viên.**
> **Context**\
The data were obtained in a survey of students math and portuguese language courses in secondary school. It contains a lot of interesting social, gender and study information about students. You can use it for some EDA or try to predict students final grade

**Nguồn thông tin và cách thu thập dư liệu:**
>- P. Cortez and A. Silva. Using Data Mining to Predict Secondary School Student Performance. In A. Brito and J. Teixeira Eds., Proceedings of 5th FUture BUsiness TEChnology Conference (FUBUTEC 2008) pp. 5-12, Porto, Portugal, April, 2008, EUROSIS, ISBN 978-9077381-39-7.
>- Fabio Pagnotta, Hossain Mohammad Amran.
>- Email:fabio.pagnotta@studenti.unicam.it, mohammadamra.hossain '@' studenti.unicam.it
>- University Of Camerino
 
**License: [CC0: Public Domain](https://creativecommons.org/publicdomain/zero/1.0/)**
>The person who associated a work with this deed has dedicated the work to the public domain by waiving all of his or her rights to the work worldwide under copyright law, including all related and neighboring rights, to the extent allowed by law. You can copy, modify, distribute and perform the work, even for commercial purposes, all without asking permission.

**Bộ dữ liệu được thu thập từ  [kaggle](https://www.kaggle.com/uciml/student-alcohol-consumption)**

---

# B. KHÁM PHÁ VÀ TIỀN XỬ LÝ DỮ LIỆU

## Đọc dữ liệu từ file .csv

In [2]:
student_por_df = pd.read_csv("./Data/student-por.csv", sep=',')
student_mat_df = pd.read_csv("./Data/student-mat.csv", sep=',')

In [3]:
student_por_df.head()

school sex  age address famsize Pstatus  Medu  Fedu     Mjob      Fjob  ...  \
0     GP   F   18       U     GT3       A     4     4  at_home   teacher  ...   
1     GP   F   17       U     GT3       T     1     1  at_home     other  ...   
2     GP   F   15       U     LE3       T     1     1  at_home     other  ...   
3     GP   F   15       U     GT3       T     4     2   health  services  ...   
4     GP   F   16       U     GT3       T     3     3    other     other  ...   

  famrel freetime  goout  Dalc  Walc health absences  G1  G2  G3  
0      4        3      4     1     1      3        4   0  11  11  
1      5        3      3     1     1      3        2   9  11  11  
2      4        3      2     2     3      3        6  12  13  12  
3      3        2      2     1     1      5        0  14  14  14  
4      4        3      2     1     2      5        0  11  13  13  

[5 rows x 33 columns]

In [4]:
student_mat_df.head()

school sex  age address famsize Pstatus  Medu  Fedu     Mjob      Fjob  ...  \
0     GP   F   18       U     GT3       A     4     4  at_home   teacher  ...   
1     GP   F   17       U     GT3       T     1     1  at_home     other  ...   
2     GP   F   15       U     LE3       T     1     1  at_home     other  ...   
3     GP   F   15       U     GT3       T     4     2   health  services  ...   
4     GP   F   16       U     GT3       T     3     3    other     other  ...   

  famrel freetime  goout  Dalc  Walc health absences  G1  G2  G3  
0      4        3      4     1     1      3        6   5   6   6  
1      5        3      3     1     1      3        4   5   5   6  
2      4        3      2     2     3      3       10   7   8  10  
3      3        2      2     1     1      5        2  15  14  15  
4      4        3      2     1     2      5        4   6  10  10  

[5 rows x 33 columns]

## Dữ liệu có bao nhiêu dòng và bao nhiêu cột?

In [5]:
n_por_rows, n_por_cols = student_por_df.shape
n_mat_rows, n_mat_cols = student_mat_df.shape
print(n_por_rows, n_por_cols)
print(n_mat_rows, n_mat_cols)

649 33
395 33


- Dữ liệu khảo sát của học sinh/sinh viên đối với môn tiếng bồ đào nha có 649 dòng và 33 cột
- Dữ liệu khảo sát của học sinh/ sinh viên đối với môn toán có 395 dòng và 33 cột

## Mỗi dòng có ý nghĩa gì? Có vấn đề các dòng có ý nghĩa khác nhau không?

- Quan sát sơ bộ dữ liệu ta thấy mỗi dòng chứa thông tin về một học sinh/ sinh viên, có vẻ như không có vấn đề, các dòng có cùng ý nghĩa với nhau.

## Dữ liệu có các dòng bị lặp không?

In [6]:
have_duplicated_por_rows= n_por_rows!= len(student_por_df.drop_duplicates())
have_duplicated_mat_rows= n_mat_rows!= len(student_mat_df.drop_duplicates())

In [7]:
print(have_duplicated_por_rows)
print(have_duplicated_mat_rows)

False
False


Vậy các dòng trong 2 tập dữ liệu không bị trùng nhau

## Mỗi cột có ý nghĩa gì?

Dưới đây là phần mô tả về  các cột trong file "student.txt":
- school - student's school (binary: 'GP' - Gabriel Pereira or 'MS' - Mousinho da Silveira)
- sex - student's sex (binary: 'F' - female or 'M' - male)
- age - student's age (numeric: from 15 to 22)
- address - student's home address type (binary: 'U' - urban or 'R' - rural)
- famsize - family size (binary: 'LE3' - less or equal to 3 or 'GT3' - greater than 3)
- Pstatus - parent's cohabitation status (binary: 'T' - living together or 'A' - apart)
- Medu - mother's education (numeric: 0 - none, 1 - primary education (4th grade), 2 – 5th to 9th grade, 3 – secondary education or 4 – higher education)
- Fedu - father's education (numeric: 0 - none, 1 - primary education (4th grade), 2 – 5th to 9th grade, 3 – secondary education or 4 – higher education)
- Mjob - mother's job (nominal: 'teacher', 'health' care related, civil 'services' (e.g. administrative or police), 'at_home' or 'other')
- Fjob - father's job (nominal: 'teacher', 'health' care related, civil 'services' (e.g. administrative or police), 'at_home' or 'other')
- reason - reason to choose this school (nominal: close to 'home', school 'reputation', 'course' preference or 'other')
- guardian - student's guardian (nominal: 'mother', 'father' or 'other')
- traveltime - home to school travel time (numeric: 1 - <15 min., 2 - 15 to 30 min., 3 - 30 min. to 1 hour, or 4 - >1 hour)
- studytime - weekly study time (numeric: 1 - <2 hours, 2 - 2 to 5 hours, 3 - 5 to 10 hours, or 4 - >10 hours)
- failures - number of past class failures (numeric: n if 1<=n<3, else 4)
- schoolsup - extra educational support (binary: yes or no)
- famsup - family educational support (binary: yes or no)
- paid - extra paid classes within the course subject (Math or Portuguese) (binary: yes or no)
- activities - extra-curricular activities (binary: yes or no)
- nursery - attended nursery school (binary: yes or no)
- higher - wants to take higher education (binary: yes or no)
- internet - Internet access at home (binary: yes or no)
- romantic - with a romantic relationship (binary: yes or no)
- famrel - quality of family relationships (numeric: from 1 - very bad to 5 - excellent)
- freetime - free time after school (numeric: from 1 - very low to 5 - very high)
- goout - going out with friends (numeric: from 1 - very low to 5 - very high)
- Dalc - workday alcohol consumption (numeric: from 1 - very low to 5 - very high)
- Walc - weekend alcohol consumption (numeric: from 1 - very low to 5 - very high)
- health - current health status (numeric: from 1 - very bad to 5 - very good)
- absences - number of school absences (numeric: from 0 to 93)
- G1 - first period grade (numeric: from 0 to 20)
- G2 - second period grade (numeric: from 0 to 20)
- G3 - final grade (numeric: from 0 to 20, output target)


## Mỗi cột hiện đang có kiểu dữ liệu gì? Có cột nào có kiểu dữ liệu chưa phù hợp để có thể xử lý tiếp không?

Để thuận tiện cho việc xử lí bên dưới,ta tiền hành gộp 2 dataframe `student_por_df`, `student_mat_df` lại thành dataframe mới có tên là `student_df` và thêm cột `subject` để phân biệt

In [8]:
new_col = pd.DataFrame(
    np.zeros(len(student_por_df)),
    columns=["subject"])["subject"].replace(0,"P").to_frame()
new_student_por_df = pd.concat([student_por_df, new_col], axis=1)
new_col = pd.DataFrame(
    np.zeros(len(student_mat_df)),
    columns=["subject"])["subject"].replace(0,"M").to_frame()
new_student_mat_df = pd.concat([student_mat_df, new_col], axis=1)
student_df = pd.concat([new_student_por_df, new_student_mat_df],
                       ignore_index=True, 
                       axis=0)

In [9]:
student_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1044 entries, 0 to 1043
Data columns (total 34 columns):
 #   Column      Non-Null Count  Dtype 
---  ------      --------------  ----- 
 0   school      1044 non-null   object
 1   sex         1044 non-null   object
 2   age         1044 non-null   int64 
 3   address     1044 non-null   object
 4   famsize     1044 non-null   object
 5   Pstatus     1044 non-null   object
 6   Medu        1044 non-null   int64 
 7   Fedu        1044 non-null   int64 
 8   Mjob        1044 non-null   object
 9   Fjob        1044 non-null   object
 10  reason      1044 non-null   object
 11  guardian    1044 non-null   object
 12  traveltime  1044 non-null   int64 
 13  studytime   1044 non-null   int64 
 14  failures    1044 non-null   int64 
 15  schoolsup   1044 non-null   object
 16  famsup      1044 non-null   object
 17  paid        1044 non-null   object
 18  activities  1044 non-null   object
 19  nursery     1044 non-null   object
 20  higher  

Đọc phần mô tả các cột,ta thấy được các cột `Medu`, `Fedu`, `traveltime`, `studytime`, `failures`, `famrel`, `freetime`, `goout`, `Dalc`, `Walc`, `health` là dạng phân loại nhưng được biểu diễn dưới dạng số. Vì vậy ta cần chuyển các cột này vè dạng phân loại

## Tiền xử lý

In [10]:
def edu (s):
    return s.replace({
        0: 'None',
        1: 'primary education',
        2: '5th to 9th grade',
        3: 'secondary education',
        4: 'higher education'
    })
student_df["Medu"] = student_df["Medu"].agg(edu)
student_df["Fedu"] = student_df["Fedu"].agg(edu)
student_df["traveltime"] = student_df["traveltime"].replace({
    1: '<15 min.',
    2: '15 to 30 min.',
    3: '30 min. to 1 hour',
    4: '>1 hour'
})
student_df["studytime"] = student_df["studytime"].replace({
    1: '1',
    2: '2 to 5 hours',
    3: '5 to 10 hours',
    4: '>10 hours'
})
student_df["failures"] = student_df["failures"].replace({
    0: '0',
    1: '1',
    2: '2',
    3: '>=3'
})
def quality(s):
    return s.replace({
    1: 'very bad',
    2: 'bad',
    3: 'normal',
    4: 'good',
    5: 'very good'
})
def quantity(s):
        return s.replace({
    1: 'very low',
    2: 'low',
    3: 'normal',
    4: 'high',
    5: 'very high'
})
student_df["famrel"] = student_df["famrel"].agg(quality)
student_df["freetime"] = student_df["freetime"].agg(quantity)
student_df["goout"] = student_df["goout"].agg(quantity)
student_df["Dalc"] = student_df["Dalc"].agg(quantity)
student_df["Walc"] = student_df["Walc"].agg(quantity)
student_df["health"] = student_df["health"].agg(quality)

In [11]:
student_df.sample(10)

school sex  age address famsize Pstatus                 Medu  \
1028     MS   F   17       R     GT3       T  secondary education   
925      GP   F   18       R     GT3       A  secondary education   
1002     MS   M   19       R     GT3       T    primary education   
153      GP   F   15       U     GT3       T     higher education   
391      GP   F   17       U     GT3       T  secondary education   
220      GP   F   16       U     GT3       A  secondary education   
1034     MS   F   18       R     GT3       T     5th to 9th grade   
215      GP   F   16       U     GT3       T     5th to 9th grade   
282      GP   M   18       U     GT3       T  secondary education   
138      GP   M   15       R     GT3       T  secondary education   

                     Fedu      Mjob      Fjob  ...   freetime      goout  \
1028    primary education   at_home     other  ...  very high       high   
925      5th to 9th grade     other  services  ...   very low   very low   
1002    primary education     other     other  ...       high       high   
153      higher education   teacher  services  ...        low        low   
391      higher education  services     other  ...       high  very high   
220     primary education  services     other  ...     normal     normal   
1034     5th to 9th grade   at_home     other  ...     normal     normal   
215   secondary education     other     other  ...       high     normal   
282      5th to 9th grade  services     other  ...       high  very high   
138      higher education   at_home   teacher  ...     normal     normal   

          Dalc      Walc     health absences  G1  G2  G3 subject  
1028       low    normal   very bad       17  10  10  10       M  
925   very low  very low  very good       75  10   9   9       M  
1002    normal    normal  very good        4   8   8   8       M  
153   very low  very low  very good        2  13  13  13       P  
391   very low    normal  very good        8  11  13  14       P  
220        low       low       good        2  11  11  12       P  
1034  very low    normal       good        2  10   9  10       M  
215   very low    normal       good        4  11  12  12       P  
282        low      high  very good        8   7   8   7       P  
138   very low  very low  very good        2  12  11  11       P  

[10 rows x 34 columns]

---

## Phân bố dữ liệu

### Với mỗi cột có kiểu dữ liệu dạng số (numerical), các giá trị được phân bố như thế nào?

In [12]:
def missing_ratio(s):
    return s.isna().mean() * 100
student_df.loc[:,student_df.dtypes == np.int64].agg([
    missing_ratio, 
    min,
    max,
    pd.Series.nunique
])

age  absences    G1    G2    G3
missing_ratio   0.0       0.0   0.0   0.0   0.0
min            15.0       0.0   0.0   0.0   0.0
max            22.0      75.0  19.0  19.0  20.0
nunique         8.0      35.0  18.0  17.0  19.0

### Với mỗi cột có kiểu dữ liệu dạng phân loại  (categorical), các giá trị được phân bố như thế nào?

In [13]:
def missing_ratio(s):
    return s.isna().mean() * 100
def num_diff_vals(s):
    return s.dropna().nunique()
def diff_vals(s):
    return s.dropna().unique()
def most_frequent(s):
    return s.value_counts().idxmax()
def most_infrequent(s):
    return s.value_counts().idxmin()
student_df.loc[:,student_df.dtypes != np.int64].agg([
    missing_ratio, 
    num_diff_vals,
    diff_vals,
    most_frequent,
    most_infrequent,
])

school     sex address     famsize Pstatus  \
missing_ratio         0.0     0.0     0.0         0.0     0.0   
num_diff_vals           2       2       2           2       2   
diff_vals        [GP, MS]  [F, M]  [U, R]  [GT3, LE3]  [A, T]   
most_frequent          GP       F       U         GT3       T   
most_infrequent        MS       M       R         LE3       A   

                                                              Medu  \
missing_ratio                                                  0.0   
num_diff_vals                                                    5   
diff_vals        [higher education, primary education, secondar...   
most_frequent                                     higher education   
most_infrequent                                               None   

                                                              Fedu  \
missing_ratio                                                  0.0   
num_diff_vals                                                    5   
diff_vals        [higher education, primary education, 5th to 9...   
most_frequent                                     5th to 9th grade   
most_infrequent                                               None   

                                                        Mjob  \
missing_ratio                                            0.0   
num_diff_vals                                              5   
diff_vals        [at_home, health, other, services, teacher]   
most_frequent                                          other   
most_infrequent                                       health   

                                                        Fjob  \
missing_ratio                                            0.0   
num_diff_vals                                              5   
diff_vals        [teacher, other, services, health, at_home]   
most_frequent                                          other   
most_infrequent                                       health   

                                            reason  ...     higher   internet  \
missing_ratio                                  0.0  ...        0.0        0.0   
num_diff_vals                                    4  ...          2          2   
diff_vals        [course, other, home, reputation]  ...  [yes, no]  [no, yes]   
most_frequent                               course  ...        yes        yes   
most_infrequent                              other  ...         no         no   

                  romantic                                    famrel  \
missing_ratio          0.0                                       0.0   
num_diff_vals            2                                         5   
diff_vals        [no, yes]  [good, very good, normal, very bad, bad]   
most_frequent           no                                      good   
most_infrequent        yes                                  very bad   

                                                 freetime  \
missing_ratio                                         0.0   
num_diff_vals                                           5   
diff_vals        [normal, low, high, very low, very high]   
most_frequent                                      normal   
most_infrequent                                  very low   

                                                    goout  \
missing_ratio                                         0.0   
num_diff_vals                                           5   
diff_vals        [high, normal, low, very low, very high]   
most_frequent                                      normal   
most_infrequent                                  very low   

                                                     Dalc  \
missing_ratio                                         0.0   
num_diff_vals                                           5   
diff_vals        [very low, low, very high, normal, high]   
most_frequent                                    very low   
most_infrequent                                      high

In [14]:
student_df["Medu"].value_counts().idxmin()
# student_df["Medu"].min()

'None'

---

# C. Đặt câu hỏi 

## Tiền xử lí và phân tích dữ liệu để trả lời cho từng câu hỏi